### ISSUE TITLE
Valet Park dialplan bug

### ISSUE THREAD

**Issue by nyfon** [fusionpbx/fusionpbx#1830]
The "park_in_use=${regex ${valet_info park@${domain_name}}|${park_lot}}" doesn't work. It gives a regex compile error and always evaluates to false. 
When the parking lot is occupied and somebody tries to park another call into the same lot it bridges a parked call to transferee, which is highly undesirable. If this code would work it wold have a proper behavior by returning the call back to the transferor...
The issue as I can see is in the regex syntax. The expression ${valet_info park@${domain_name}} is not being evaluated.
Here is the proper syntax that works: park_in_use=${regex ${eval ${valet_info park@${domain_name}}}|${park_lot:1}}
Also, the parking lot has a special character (*) and it's not escaped thus giving a regex runtime error. I had to chop off the * from the lot for it to work properly.I tried to escape it inline but it didn't work. Maybe expand command will help - I'm not sure

**Comment by nyfon** [comment]
Here is the lua script which will do better than regex-ing in dialplan...
just need to call it with park_lot parameter...

	api = freeswitch.API();
	park_lot = argv[1];
	
	if not park_lot or park_lot == "" then return end;
--get active lots for the domain
	domain_name = session:getVariable("domain_name");
	context 	= session:getVariable("context");
	hold_music	= session:getVariable("hold_music");
	
	lot_in_use = string.find(api:executeString("valet_info park@" .. domain_name),park_lot) or false;
	
	referred_by_user=string.match(session:getVariable("sip_h_Referred-By") or "",'sip:(.*)@.*') or "";
	
	if not(referred_by_user=="") then
	-- if trying to park to the occupied lot return call back to the referrer
		if lot_in_use then
			session:execute("transfer", referred_by_user .." XML "..context);
			return;
		end
    else
	-- user just pressed an empty parking lot button... respond with 404
		if not lot_in_use then
			session:execute("respond", "404");
			return;
		end
	end
-- do the actual valet park stuff here
	session:setVariable("valet_parking_timeout","300");
	session:setVariable("valet_hold_music",hold_music);
	session:setVariable("valet_parking_orbit_exten",referred_by_user);
	session:execute("valet_park", "park@"..domain_name.." "..park_lot);

**Comment by jlcrews** [comment]
I agree completely this is highly undesirable.  I wish there was a way to have a way to have park or a copy of the park function modified to not bridge two callers together if you park two calls in the same lot.  We need basic key system functionality.  Having two unrelated callers bridged together in the same park is horrible.

**Comment by markjcrane** [comment]
Yealink key function call park will not park calls on top of each other. Polycom does not do this for you this is one of the reasons I recommend Yealink.

**Comment by jlcrews** [comment]
I just tried this on a Yealink ...when I had a call parked in a lot...and when I tried to park a second call in the same lot the Yealink phone said "transfer failed" and automatically place the call on hold on the Yealink phone.  So that is nice.  Going forward I will recommend Yealink...but what do I do for my customers with over 50 phones now that I just figured this out?  There has to be some way to approach this differently.

**Comment by nyfon** [comment]
Unfortunately, we cannot depend on the brand handling the parking... I have multiple brad deployments and that was a big issue. The code I've attached handles the situation properly for most of the brands...
The only case is Aastra and incoming call - that's a complete mess so I had to write a custom logic because the Aastra sends a 302 Deflection and the button type can be set only to BLF/Xfer...

**Comment by jlcrews** [comment]
Nyfon...can you share in detail how to use this code you provided in FusionPBX?  Thanks.

**Comment by nyfon** [comment]
sure - just delete everything up to the first line (first condition order# 5 ) in order# 470 (valet_park) in the dialplan manager and put action|lua|valet_park.lua $2 as order# 10. Don't forget to put the above attached script (valet_park.txt rename it to valet_park.lua) to your freeswitch scripts directory

**Comment by netpro25** [comment]
Mark I think this is important and should be merged with the code as it effects almost all non-yealink phones.
